### DISCUSSION TITLE
How to use multi-material with Klipper-macro based tool changing?

### DISCUSSION THREAD

**Discussion by goeland86** [supermerill/SuperSlicer#4157] [Q&A]
Hi,

I've set up my 3D Chameleon printer to use Klipper macros for any and all tool changing. I finished testing the macros and they do exactly what I need them to.

Now my issue is that, well, SuperSlicer seems to inject color priming code *before* the start gcode, and then tries to outsmart me in how to manage color changing.

Is there a way I can tell SuperSlicer to just use `T0`, `T1`, `T2` and `T3` as the *only* gcode needed for color swaps? I may just be blind as to how to do so, which is why I didn't open up an issue yet.

**Comment by supermerill** [comment]
put your macro call into the toolchange custom gcode, and it should be good.
as en exemple, you can put `T{next_extruder}` or `TOOLCHANGE={extruder_name}` (if you set an extruder_name for each extruder)

See tooltip for more information: it put automatically the TX if there is nothing inside it.

For the color priming, you can disable it, i guess? Depends on what you talked about.

**Comment by goeland86** [comment]
I've done that, but it seems to be adding initial purge prep before the initial stuff is happening?

This is the header right after the embedded thumbnail string:
```
; TODO: check the size of the extruder and the y-bar clearance. Optimize & test supports.

; external perimeters extrusion width = 0.84mm
; perimeters extrusion width = 0.84mm
; infill extrusion width = 0.84mm
; solid infill extrusion width = 0.84mm
; top infill extrusion width = 0.84mm
; first layer extrusion width = 0.84mm

; external perimeters extrusion width = 0.84mm
; perimeters extrusion width = 0.84mm
; infill extrusion width = 0.84mm
; solid infill extrusion width = 0.84mm
; top infill extrusion width = 0.84mm
; first layer extrusion width = 0.84mm


G1 Z0.3 F18000
;_STORE_FAN_SPEED_WT
;HEIGHT:0.3
;TYPE:Wipe tower
;WIDTH:0.75
;--------------------
; CP PRIMING START
;--------------------
M220 S100
G1  X6.000 Y0.385   F7200
; Filament-specific end gcode 
;END gcode for filament

T1
;_TOOLCHANGE 1

G1 F18000
G1  X6.000 Y0.385  
; Filament gcode

G1  X6.000 Y0.385 Z0.3
G4 S0
M220 S100
; CP TOOLCHANGE WIPE
G1  X66.000  E5.1309 F95040
G1  X64.875 Y1.135   E0.1156
G1  X6.000  E5.0347
G1  X64.875  F7200
; CP TOOLCHANGE UNLOAD
;WIDTH:0.9
G1  X6.750 Y2.335  
SET_PRESSURE_ADVANCE ADVANCE=0
G1 F1580
G1  X13.332  E0.686
G1 F1628
```

My issue here is that it's doing these extruder and gantry related moves before `print_start`, where I home and initialize the tool change setup to be ready for this. If it's a bug, I'll happily submit an issue.

**Reply by goeland86** [reply]
Attaching the .3mf project file for convenience in testing/understanding with my setup.
What I expect would be something like:
```
PRINT_START EXTRUDER_TEMP=215 BED_TEMP=65
T0
<paths for T0>

T1
<paths for T1>
```

Thanks for the help clarifying this.
[Dice_holes_project.zip](https://github.com/supermerill/SuperSlicer/files/14419515/Dice_holes_project.zip)
